### PR TITLE
add an allHelp command to show all the help in one go

### PR DESCRIPTION
**Pull request by bthaile**
References: https://github.com/onflow/docs/issues/424

## Description

Displays all the help in one go, user can pip into `grep` or `more`. Also easy to add output to developer docs.
______

For contributor use:

- [x] Targeted PR against `master` branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/flow-cli/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

File: internal/super/all_help.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	cmd := c.Root()
+	// Slice to hold the help texts of all commands
+	var helpTexts []string
+	helpTexts = append(helpTexts, "```")
+
+	// Generate the help texts
+	generateCommandHelpTexts(cmd, &helpTexts)
+	helpTexts = append(helpTexts, "```")
+
+	commands := strings.Join(helpTexts, "\n\n------------\n\n")
```
Comment (turbolent) [inline internal/super/all_help.go:39]: consider using `strings.Builder`

File: internal/super/all_help.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		} else {
+			// If there's no section after Global Flags, remove starting from the marker to the end of the text
+			helpText = helpText[:start]
+		}
+	}
+	return helpText
+}
+
+func removeANSICodes(input string) string {
+	re := regexp.MustCompile(`\x1B\[[0-9;]*[a-zA-Z]`)
```
Comment (turbolent) [inline internal/super/all_help.go:98]: Compile the regex once and reuse it. Make it a global

Comment (bthaile) [reply internal/super/all_help.go]: Thanks updated the PR.

File: internal/super/all_help.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	if removeGlobalFlags {
+		helpText = removeGlobalFlagsSection(helpText)
+	}
+
+	return ansiRegex.ReplaceAllString(helpText, "")
+}
+
+func removeGlobalFlagsSection(helpText string) string {
+	// Define the start of the Global Flags section
+	startMarker := "Global Flags:\n"
```
Comment (turbolent) [inline internal/super/all_help.go:102]:
```suggestion
	const startMarker = "Global Flags:\n"
```

File: internal/super/all_help.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	"regexp"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var ansiRegex = regexp.MustCompile(`\x1B\[[0-9;]*[a-zA-Z]`)
+
+var AllHelp = &cobra.Command{
+	Use:     "allHelp",
```
Comment (chasefleming) [inline internal/super/all_help.go:32]: This should probably be cased as `all-help` to match other commands.

**Review by chasefleming — CHANGES_REQUESTED** [review]

**Review by turbolent — APPROVED** [review]

**Comment by jribbink** [comment]
Yeah +1 to maintaining this on docs, actions PR would probably help.  Maybe this is an alternate discussion, but it feels like having this on the docs site almost a bit of a bandaid to CLI reference docs which should generally be the source of this syantax/usage information

**Comment by bthaile** [comment]
What about auto gen for flow cli and link to it from developer docs?

**Comment by jribbink** [comment]
I'm curious if there's any way we can avoid introducing a new top-level CLI command (we're pretty bloated already which is ironically what this command is trying to fix :P).  Was it explored if it were possible to add this as a flag to the `flow help` command -- i.e. `flow help --all`

File: internal/super/all_help.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	return helpTexts.String(), nil
+
+}
+
+// Recursive function to traverse all commands and subcommands,
+// capturing the help text for each.
+func generateCommandHelpTexts(cmd *cobra.Command, result *strings.Builder) {
+	removeGlobalFlags := true
+	// version is last command show global flags
+	if cmd.Name() == "version" {
```
Comment (chasefleming) [inline internal/super/all_help.go:66]: Checking the string name here is prone to break at some point. Can you check for the last item instead please?

**Review by chasefleming — CHANGES_REQUESTED** [review]

**Comment by bthaile** [comment]
I think there is a way to add "--all" to help. I'll research it.

**Comment by bthaile** [comment]
added custom help command and included "--all" or "-a" to display all help. left command all-help available as well.

**Comment by bthaile** [comment]
custom help didn't behave as expected, so I removed it.

File: internal/super/all_help.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var ansiRegex = regexp.MustCompile(`\x1B\[[0-9;]*[a-zA-Z]`)
```
Comment (jribbink) [inline internal/super/all_help.go:29]: what is this regex used for?

Comment (chasefleming) [inline internal/super/all_help.go:29]: What is this doing?

File: internal/super/all_help.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+var AllHelp = &cobra.Command{
```
Comment (chasefleming) [inline internal/super/all_help.go:31]: `AllHelp := ...`

File: internal/super/all_help.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	// Define the start of the Global Flags section
+	const startMarker = "Global Flags:\n"
+	start := strings.Index(helpText, startMarker)
+	if start != -1 {
+		// Attempt to find the end of the Global Flags section by looking for the next section or end of text
+		// This may need adjustment based on your actual help text formatting.
+		end := strings.Index(helpText[start+len(startMarker):], "\n\n")
+		if end != -1 {
+			// Remove the Global Flags section
+			helpText = helpText[:start] + helpText[start+end+len(startMarker)+2:]
```
Comment (chasefleming) [inline internal/super/all_help.go:101]: Does this work 100% of the time in this case?

File: internal/super/all_help.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		fmt.Printf("Error generating help for %s", cmd.Name())
+	}
+
+	cmd.SetOut(nil)
+	helpText := sb.String()
+	helpText = removeGlobalFlagsSection(helpText)
+	return ansiRegex.ReplaceAllString(helpText, "")
+}
+
+func removeGlobalFlagsSection(helpText string) string {
```
Comment (chasefleming) [inline internal/super/all_help.go:91]: How reliable is this?

File: internal/super/all_help.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+	"github.com/spf13/cobra"
+)
+
+var ansiRegex = regexp.MustCompile(`\x1B\[[0-9;]*[a-zA-Z]`)
+
+var AllHelp = &cobra.Command{
+	Use:     "all-help",
+	Short:   "Outputs help for all the CLI commands",
+	Example: "flow cheatsheet",
```
Comment (chasefleming) [inline internal/super/all_help.go:34]: The example does not match the command

File: internal/super/all_help.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+	Run: func(
+		c *cobra.Command,
+		args []string,
+	) {
+		root := c.Root()
+		r, err := generateCS(root)
+		if err != nil {
+			fmt.Printf("Error generating cheatsheet: %s", root.Name())
```
Comment (chasefleming) [inline internal/super/all_help.go:42]: Would change the wording from `cheatsheet` to match the command more.

File: internal/super/all_help.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+	return helpTexts.String(), nil
+
+}
+
+// Recursive function to traverse all commands and subcommands,
+// capturing the help text for each.
+func generateCommandHelpTexts(cmd *cobra.Command, result *strings.Builder) {
+
+	result.WriteString("\n\n---------------\n\n")
```
Comment (chasefleming) [inline internal/super/all_help.go:66]: I would avoid writing this twice and move it somewhere to be shared.

**Review by chasefleming — CHANGES_REQUESTED** [review]

**Comment by chasefleming** [comment]
@bthaile I'm wondering if this is already solved for you by Cobra somehow. Have you seen this? https://pkg.go.dev/github.com/spf13/cobra/doc#GenMarkdownTree

**Comment by bthaile** [comment]
closing, doesn't seem to fit in the command motif

**Pull request closed without merging**